libr/socket/socket.c

```diff
@@ -301,16 +301,40 @@ R_API bool r_socket_connect(RSocket *s, const char *host, const char *port, int
 				perror ("socket");
 				continue;
 			}
-			ret = setsockopt (s->fd, IPPROTO_TCP, TCP_NODELAY, &flag, sizeof (flag));
-			if (ret < 0) {
-				perror ("setsockopt");
-				close (s->fd);
-				s->fd = -1;
-				continue;
+			if (proto == R_SOCKET_PROTO_TCP) {
+				ret = setsockopt (s->fd, IPPROTO_TCP, TCP_NODELAY, &flag, sizeof (flag));
+				if (ret < 0) {
+					perror ("setsockopt");
+					close (s->fd);
+					s->fd = -1;
+					continue;
+				}
 			}
 
-			r_socket_block_time (s, 0, 0, 0);
-			ret = connect (s->fd, rp->ai_addr, rp->ai_addrlen);
+			// Bind the source port to the same dest port
+			if (proto == R_SOCKET_PROTO_UDP) {
+				memset (&s->sa, 0, sizeof (s->sa));
+				s->sa.sin_family = AF_INET;
+				s->sa.sin_addr.s_addr = htonl (s->local? INADDR_LOOPBACK: INADDR_ANY);
+				s->port = r_socket_port_by_name (port);
+				if (s->port < 1) {
+					return false;
+				}
+				s->sa.sin_port = htons (s->port); // TODO honor etc/services
+				if (bind (s->fd, (struct sockaddr *)&s->sa, sizeof (s->sa)) < 0) {
+					r_sys_perror ("bind");
+#ifdef _MSC_VER
```

Comment:
why not use `__WINDOWS__` ?

Reply:
No reason as I followed some code in `r_socket_listen`. I can change it.